homeassistant/components/mqtt/light/schema_json.py

```diff
@@ -463,7 +463,10 @@ def supported_features(self):
                 message["flash"] = self._flash_times[CONF_FLASH_TIME_SHORT]
 
         if ATTR_TRANSITION in kwargs:
-            message["transition"] = int(kwargs[ATTR_TRANSITION])
+            if kwargs[ATTR_TRANSITION].is_integer():
```

Comment:
The light transition will be validated as a float, in the service schema preceding the actuator method call.

https://github.com/home-assistant/home-assistant/blob/6cc71db385ce9c9cf9bac32624568ee5658b92b1/homeassistant/components/light/__init__.py#L93

Reply:
So the idea is to send an integer if the float has 0 as decimal and otherwise send a float?

Reply:
@MartinHjelmare yes kwargs[ATTR_TRANSITION] is already a float.
However the current code typcasts that float to an int.
the .is_integer() works on a float to determine if that float is an "whole number" if that is the case it will convert the float to an int.
Therefore if you call a transition with value 2 for instance, this will ensure that the MQTT message contains "2" instead of "2.0".

I think in most cases this will not matter, but I thought there might be some light out there in the world that cannot cope with floating points and would get upset with a value of "2.0" instead of "2".

Therefore I wanted to make it backwords compatible.

Reply:
@MartinHjelmare your second post is completly correct.
Sorry i did not read that before typing my replay

Reply:
Would be nice if we can always send the same type of value, but I don't know the device landscape very well.

Reply:
I could just change it to always send a float?
for all my devices that would be fine.
I am just afraid I might brake it for someone else who uses some diffrent MQTT light...

Reply:
@emontnemery what do you think?

Reply:
According to the fine docs, multiple implementations are support MQTT light with JSON schema: https://www.home-assistant.io/integrations/light.mqtt/#implementations

Based on a quick code review, the implementations which support transitions are using ArduinoJson to parse the JSON and should be fine:

- AiLight supports only integer transitions:
https://github.com/stelgenhof/AiLight/blob/8f937e23c1d49a663f8ddafbe63b073dc51fbc83/src/light.ino#L166

- ESPHome supports floats:
https://github.com/esphome/esphome/blob/4bc306772590b6b6df5bf5e28ea356d17e8a272f/esphome/components/light/light_state.cpp#L244

- esp-mqtt-rgb-led supports only integer transitions:
https://github.com/corbanmailloux/esp-mqtt-rgb-led/blob/9d45df2f4c4dfcfd169e0f2b3d04b47df91921a5/mqtt_esp8266_light/mqtt_esp8266_light.ino#L297

@starkillerOG What software are you using which supports MQTT light with JSON schema?

Reply:
@emontnemery I actually wrote my own software and put it on github if other people might be instrested: https://github.com/starkillerOG/h801-mqtt-json
It is meant for a H801 LED controller (about €10,- on aliexpress, inside is an ESP8266 for the wifi) that supports 12V RGB+cold-white+warm-white LED strips.
I wrote my own code because I could not find any code at the time that supported MQTT json that implemented color-temperature to control the warm/cold white LED strip together with the RGB strip as a single light in HomeAssistant.

Reply:
@emontnemery I also use ArduinoJson to parse the JSON by the way.
If you want I could put a link to my github project in the implementations section of the HomeAssistant docs that you refered to: https://www.home-assistant.io/integrations/light.mqtt/#implementations

@emontnemery, @MartinHjelmare, Schould I change the code to always sent a float or keep it like this to sent a int if it is a "whole number" and otherwise a float?

Reply:
@starkillerOG I think it's fine to change the code to always send a float.

But can you please do some testing on your side to verify ArduinoJson is working as expected?

Reply:
will test that and report back soon.

Reply:
@emontnemery I just tested the `int transitionTime = root["transition"];`.
It works fine if you sent a float, it just floors the float so 2.8 --> 2 
Therefore I will update this PR to just always sent a float.